okhttp-backend/src/main/scala/sttp/client/okhttp/OkHttpBackend.scala

```diff
@@ -79,117 +68,58 @@ abstract class OkHttpBackend[F[_], S <: Streams[S], P](
-        Some(OkHttpRequestBody.create(b, mediaType))
-      case ByteBufferBody(b, _) =>
-        if (b.isReadOnly) Some(OkHttpRequestBody.create(ByteString.of(b), mediaType))
-        else Some(OkHttpRequestBody.create(b.array(), mediaType))
-      case InputStreamBody(b, _) =>
-        Some(new OkHttpRequestBody() {
-          override def writeTo(sink: BufferedSink): Unit =
-            sink.writeAll(Okio.source(b))
-          override def contentType(): MediaType = mediaType
-        })
-      case FileBody(b, _) =>
-        Some(OkHttpRequestBody.create(b.toFile, mediaType))
-      case StreamBody(s) =>
-        streamToRequestBody(s.asInstanceOf[streams.BinaryStream])
-      case MultipartBody(ps) =>
-        val b = new OkHttpMultipartBody.Builder().setType(Option(mediaType).getOrElse(OkHttpMultipartBody.FORM))
-        ps.foreach(addMultipart(b, _))
-        Some(b.build())
-    }
-  }
-
-  private def addMultipart(builder: OkHttpMultipartBody.Builder, mp: Part[BasicRequestBody]): Unit = {
-    val allHeaders = mp.headers :+ Header(HeaderNames.ContentDisposition, mp.contentDispositionHeaderValue)
-    val headers =
-      OkHttpHeaders.of(allHeaders.filterNot(_.is(HeaderNames.ContentType)).map(h => (h.name, h.value)).toMap.asJava)
-
-    bodyToOkHttp(mp.body, mp.contentType).foreach(builder.addPart(headers, _))
-  }
+  protected val bodyToOkHttp: BodyToOkHttp[F, S]
+  protected val bodyFromOkHttp: BodyFromOkHttp[F, S]
 
   private[okhttp] def readResponse[T, R >: PE](
       res: OkHttpResponse,
       responseAs: ResponseAs[T, R]
   ): F[Response[T]] = {
-    val headers = res
-      .headers()
-      .names()
-      .asScala
-      .flatMap(name => res.headers().values(name).asScala.map(Header(name, _)))
-      .toList
-
+    val headers = readHeaders(res)
     val responseMetadata = ResponseMetadata(headers, StatusCode(res.code()), res.message())
     val encoding = headers.collectFirst { case h if h.is(HeaderNames.ContentEncoding) => h.value }
     val method = Method(res.request().method())
     val byteBody = if (method != Method.HEAD) {
       encoding
         .map(e =>
-          customEncodingHandler //There is not PartialFunction.fromFunction in scala 2.12
+          customEncodingHandler //There is no PartialFunction.fromFunction in scala 2.12
             .orElse(EncodingHandler(standardEncoding))(res.body().byteStream() -> e)
         )
         .getOrElse(res.body().byteStream())
     } else {
       res.body().byteStream()
     }
 
-    val body = responseHandler(byteBody).handle(responseAs, responseMonad, responseMetadata)
+    val body = bodyFromOkHttp(byteBody, responseAs, responseMetadata, None)
     responseMonad.map(body)(Response(_, StatusCode(res.code()), res.message(), headers, Nil))
   }
 
+  private def readHeaders[R >: PE, T](res: OkHttpResponse) = {
+    res
+      .headers()
+      .names()
+      .asScala
+      .flatMap(name => res.headers().values(name).asScala.map(Header(name, _)))
```

Comment:
maybe the scala-java conversion of headers could happen once, instead of for every header